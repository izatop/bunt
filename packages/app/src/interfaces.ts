import {Action, Context, MatchContext, Promisify} from "@typesafeunit/unit";
import {RouteAbstract} from "./Route";

export type RouteResponse = Error
    | {stringify(): string}
    | NodeJS.ReadableStream
    | Buffer
    | string
    | number
    | boolean
    | null
    | undefined
    | void;

export type RouteAction = Action<any, any, RouteResponse>;

export type MatchRoute<C extends Context, R> = R extends RouteAbstract<infer A>
    ? A extends Action<infer AC, any, RouteResponse>
        ? MatchContext<C, AC> extends AC
            ? R
            : ["Action context doesn't match its parent context", C, AC]
        : ["Action doesn't satisfy its interface", A]
    : never;


export interface IKeyValueMap {
    has(name: string): boolean;

    get(name: string, defaultValue?: string): string;

    set(name: string, value: string): void;

    delete(name: string): void;

    entries(): [string, string][];

    toObject(): { [key: string]: string };
}

export interface IKeyValueReadonlyMap extends IKeyValueMap {
    set: never;
    delete: never;
}

export interface IRequestBodyTransform<T> {
    transform(request: IRequest): Promise<T>;
}

export type RequestTransformType<T> = IRequestBodyTransform<T> | ((request: IRequest) => Promise<T>);

export interface IRequest {
    readonly route: string;
    readonly headers: IHeaders;

    getBuffer(): Promise<Buffer>;
    createReadableStream(): Promisify<NodeJS.ReadableStream>;
    transform<T>(transformer: RequestTransformType<T>): Promise<T>;
}

export type HeaderAssertValue = |
    string |
    string[] |
    ((value: string) => boolean | void);

export interface IHeaders extends IKeyValueReadonlyMap {
    assert(header: string, values: HeaderAssertValue): void;
}
