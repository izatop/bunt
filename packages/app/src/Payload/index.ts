export * from "./Resolver.js";
export * from "./Payload.js";
export * from "./interfaces.js";
