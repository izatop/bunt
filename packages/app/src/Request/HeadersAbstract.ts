import {assert, ILogable, isBoolean, isFunction} from "@bunt/util";
import {HeaderAssertValue, IHeaders} from "../interfaces.js";
import {StrictKeyValueMap} from "./StrictKeyValueMap.js";

export abstract class HeadersAbstract extends StrictKeyValueMap
    implements IHeaders, ILogable<{[key: string]: string}> {

    public assert(header: string, expected: HeaderAssertValue): void {
        const clientValue = this.get(header.toLowerCase());
        if (Array.isArray(expected)) {
            assert(
                expected.some((e) => clientValue.includes(e)),
                `Wrong header "${header}" value, allowed: ${expected.join(", ")}`,
            );
        }

        if (isFunction(expected)) {
            const result = expected(clientValue);
            if (isBoolean(result)) {
                assert(result, `Wrong ${header}`);
            }
        }
    }

    public getLogValue(): Record<any, any> {
        return this.toJSON();
    }
}
