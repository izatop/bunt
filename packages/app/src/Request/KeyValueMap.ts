import {IKeyValueMap} from "../interfaces.js";

export class KeyValueMap implements IKeyValueMap {
    readonly #map: Map<string, string>;

    constructor(values: [string, string][]) {
        this.#map = new Map(values);
    }

    public static fromObject(input: Record<string, string>): IKeyValueMap {
        return new this(Object.entries(input));
    }

    public delete(name: string): void {
        this.#map.delete(name);
    }

    public entries(): [string, string][] {
        return [...this.#map.entries()];
    }

    public has(name: string): boolean {
        return this.#map.has(name);
    }

    public get(name: string, defaultValue?: string): string {
        return this.#map.get(name) || defaultValue || "";
    }

    public toJSON(): {[p: string]: string} {
        const object: {[p: string]: string} = {};
        for (const [key, value] of this.#map.entries()) {
            object[key] = value;
        }

        return object;
    }

    public set(name: string, value: string): void {
        this.#map.set(name, value);
    }
}
