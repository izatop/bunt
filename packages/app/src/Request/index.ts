export * from "./KeyValueMap.js";
export * from "./HeadersAbstract.js";
export * from "./RequestAbstract.js";
export * from "./RequestMessageAbstract.js";
export * from "./RequestValidatorAbstract.js";
export * from "./StrictKeyValueMap.js";
