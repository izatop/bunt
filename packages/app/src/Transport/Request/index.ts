export * from "./URLEncodedFormTransform";
export * from "./JSONTransform";
