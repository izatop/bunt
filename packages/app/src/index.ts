export * from "./Payload/index.js";
export * from "./Route/index.js";
export * from "./Request/index.js";
export * from "./TransformRequest/index.js";
export * from "./Application.js";
export * from "./interfaces.js";
