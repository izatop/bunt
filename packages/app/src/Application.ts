import {Context, ContextArg, unit, Unit} from "@typesafeunit/unit";
import {assert, isFunction, isObject} from "@typesafeunit/util";
import {IRequest, MatchRoute, RouteAction, RouteResponse} from "./interfaces";
import {RouteAbstract} from "./Route";
import {TransportAbstract} from "./Transport";

export class Application<U extends Unit<C>, C> {
    protected readonly unit: U;
    protected readonly route: RouteAbstract<RouteAction>[] = [];

    constructor(u: U) {
        this.unit = u;
    }

    public get size() {
        return this.route.length;
    }

    public static async factory<C extends Context>(context: ContextArg<C>) {
        return new this<Unit<C>, C>(await unit(context));
    }

    public add<R extends RouteAbstract<RouteAction>>(route: MatchRoute<C, R>) {
        assert(!this.unit.has(route.action), `This route was already added`);
        this.unit.add(route.action);
        this.route.push(route);
        return this;
    }

    public remove<R extends RouteAbstract<RouteAction>>(route: MatchRoute<C, R>) {
        if (this.unit.has(route.action)) {
            this.unit.remove(route.action);
            const index = this.route.findIndex((item) => item === route);
            this.route.splice(index, index + 1);
        }

        return this;
    }

    public async handle(transport: TransportAbstract) {
        try {
            await transport.respond(await this.run(transport.request));
        } catch (error) {
            if (!transport.sent) {
                await transport.respond({code: 500, status: error.message});
            }

            throw error;
        }
    }

    public async run(request: IRequest): Promise<RouteResponse> {
        for (const route of this.route) {
            if (route.test(request.route)) {
                const state = {};
                const requestArgs = new Map<string, string>();
                const context = await this.unit.getContext();
                Object.entries(route.match(request.route))
                    .forEach(([key, value]) => requestArgs.set(key, value));

                const routeContext = {request, context, args: requestArgs};
                if (isFunction(route.validate)) {
                    await route.validate(routeContext);
                }

                if (isObject(route.state)) {
                    for (const [name, factory] of Object.entries(route.state)) {
                        Reflect.set(state, name, await factory(routeContext));
                    }
                }

                return this.unit.run(route.action, state);
            }
        }

        return {code: 404, status: "Page not found"};
    }
}
