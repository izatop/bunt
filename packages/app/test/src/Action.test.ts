import HelloWorldRoute from "../../../test/src/actions/HelloWorldRoute";
import {MainContext} from "../../../test/src/context/MainContext";
import {Request} from "../../../test/src/transport/Request";
import {Application, RouteNotFound} from "../../src";

describe("Route", () => {
    const headers = {"Content-Type": "application/json"};

    test("Success", async () => {
        const app = await Application.factory(new MainContext(), [HelloWorldRoute]);
        const request = new Request(
            "/u/123",
            headers,
            JSON.stringify({name: "World"}),
        );

        try {
            const response = await app.run(request);
            expect({request, response}).toMatchSnapshot();
        } catch (error) {
            console.log(error.toSafeJSON());
        }
    });

    test("Fails", async () => {
        const app = await Application.factory(new MainContext(), [HelloWorldRoute]);
        const wrongRequest = new Request(
            "/u/123",
            headers,
            JSON.stringify({}),
        );

        await expect(app.run(wrongRequest))
            .rejects
            .toThrow("Assertion failed");
    });

    test("Not found", async () => {
        const app = await Application.factory(new MainContext());
        const request = new Request("/wrong-uri", {});
        await expect(app.run(request)).rejects.toThrowError(RouteNotFound);
    });
});
