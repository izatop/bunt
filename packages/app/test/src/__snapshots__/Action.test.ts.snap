// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route Success 1`] = `
Object {
  "request": Request {
    "body": "{\\"name\\":\\"World\\"}",
    "headers": Object {
      "content-type": "application/json",
    },
    "route": "/u/123",
  },
  "response": "Hello, World!",
}
`;
